tests/test_cxl_device_multi_logical_device.py

```diff
@@ -0,0 +1,448 @@
+        packet = CxlIoMemWrPacket.create(
+            memory_base_address + bar_size, 4, data=data, ld_id=target_ld_id
+        )
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+
+        # NOTE: Write OOB (Lower Boundary), Expect No Error
+        packet = CxlIoMemWrPacket.create(memory_base_address - 4, 4, data=data, ld_id=target_ld_id)
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+
+        # NOTE: Read OOB (Upper Boundary), Expect 0
+        packet = CxlIoMemRdPacket.create(memory_base_address + bar_size, 4, ld_id=target_ld_id)
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+        packet = await packet_reader.get_packet()
+        assert is_cxl_io_completion_status_sc(packet)
+        cpld_packet = cast(CxlIoCompletionWithDataPacket, packet)
+        assert cpld_packet.data == 0
+
+        # NOTE: Read OOB (Lower Boundary), Expect 0
+        packet = CxlIoMemRdPacket.create(memory_base_address - 4, 4, ld_id=target_ld_id)
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+        packet = await packet_reader.get_packet()
+        assert is_cxl_io_completion_status_sc(packet)
+        cpld_packet = cast(CxlIoCompletionWithDataPacket, packet)
+        assert cpld_packet.data == 0
+
+    async def send_packets(
+        target_ld_id: int, reader: asyncio.StreamReader, writer: asyncio.StreamWriter
+    ):
+        packet_reader = PacketReader(reader, label="send_packets")
+        packet_writer = writer
+
+        target_address = 0x80
+        target_data = 0xDEADBEEF
+
+        logger.info("[PyTest] Sending CXL.mem request packets from client")
+        packet = CxlMemMemWrPacket.create(target_address, target_data, ld_id=target_ld_id)
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+
+        # Get ACK
+        packet = await packet_reader.get_packet()
+        # TODO: Check if the ACK is correctly formed
+
+        packet = CxlMemMemRdPacket.create(target_address, ld_id=target_ld_id)
+        packet_writer.write(bytes(packet))
+        await packet_writer.drain()
+
+        logger.info("[PyTest] Checking CXL.mem request packets received from server")
+        packet = await packet_reader.get_packet()
+        mem_packet = cast(CxlMemMemRdPacket, packet)
+        logger.info(f"[PyTest] Received CXL.mem packet: {hex(mem_packet.data)}")
+        assert mem_packet.data == target_data
+
+        # Check MLD bin file
+        logger.info("[PyTest] Checking MLD bin file")
+        with open(f"mld_mem{target_ld_id}.bin", "rb") as f:
+            f.seek(target_address)
+            data = f.read(4)
+            value = struct.unpack("<I", data)[0]
+            assert value == target_data
+
+    async def wait_test_stop():
+        # pylint: disable=protected-access
+        await gather(*(device.wait_for_ready() for device in mld._cxl_type3_devices))
```

Comment:
Do not wait for all type-3 devices to be ready, but instead wait for MLD ready which would imply all type-3 devices ready.

Reply:
Avoid iterating through internal member list.